# Biodiversity Project

This project is part of the Codecademy course "Data Scientist: Natural Language Processing Career Path".

For this project, data from National Parks Service about endangered species in different parks will be interpreted in order to check any patterns to endangered species.

### Goals

The aim of the project is to investigate any patterns or themes to the types of species that become endangered and then assess the likeliness of becoming extinct.

### Resources

Data from National Parks Service:

- observations.csv

- species_info.csv

For classification status:

- nps.gov

- fisheries.noaa.gov

- wikipedia.org

- maine.gov

Other:

- wikipedia.org

### Analysis


Some analysis of the data will intend to answer the following questions:

- What is the distribution of `conservation_status` for animals?
- Are certain types of species more likely to be endangered?
- Are the differences between species and their conservation status significant?
- Which species were spotted the most at each park?



## 1. Importing Python Libraries

This project will be using Seaborn and Matplotlib libraries to plot data. Also, this project will be using Pandas and Numpy for data manipulation.


In [230]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

## 2. Loading the data


In [231]:
species = pd.read_csv("species_info.csv")
observations = pd.read_csv("observations.csv")

## 3. Distribution of the conservation status 

Conservation status values are in the `species` dataset.

### 3.1 Inspecting the `species` dataset

In the `species` dataset there are four columns: `category`,`scientific_name`,`common_names`, and `conservation_status`.

The first five entries of the `conservation_status` columns have no data.

In [232]:
print(species.columns)
print(species.head(10))


Index(['category', 'scientific_name', 'common_names', 'conservation_status'], dtype='object')
  category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   
3   Mammal                     Ovis aries   
4   Mammal                 Cervus elaphus   
5   Mammal         Odocoileus virginianus   
6   Mammal                     Sus scrofa   
7   Mammal                  Canis latrans   
8   Mammal                    Canis lupus   
9   Mammal                    Canis rufus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN  
3  Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)                 NaN  
4                                  

As seen below, most entries in the conservation status columns (about 97%) have missing data.

In [233]:
print(species.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB
None


In [234]:
species.describe(include='all')

category    scientific_name        common_names  \
count             5824               5824                5824   
unique               7               5541                5504   
top     Vascular Plant  Castor canadensis  Brachythecium Moss   
freq              4470                  3                   7   

       conservation_status  
count                  191  
unique                   4  
top     Species of Concern  
freq                   161

### 3.1.1 Handling missing data

Since there are 191 non-null values in the `conservation_status` column, it is crucial to tackle missing data. Despite the fact that there are 5824 rows, most of them have missing data for the conservation_status column.

In [235]:
species.conservation_status.isna().value_counts()

True     5633
False     191
Name: conservation_status, dtype: int64

In [236]:
species.conservation_status.value_counts()

Species of Concern    161
Endangered             16
Threatened             10
In Recovery             4
Name: conservation_status, dtype: int64

In [237]:
# getting to know species that have a value for the conservation_status column
species_conservation = species[species.conservation_status.notna()]
species_conservation = species_conservation.rename(columns={'scientific_name': 'scientific', 'common_names': 'nicknames',\
                                     'conservation_status': 'status'})
print(species_conservation.scientific.nunique())
species_conservation.scientific.unique()


179


array(['Canis latrans', 'Canis lupus', 'Canis rufus', 'Eptesicus fuscus',
       'Lasionycteris noctivagans', 'Myotis austroriparius',
       'Myotis grisescens', 'Myotis keenii', 'Myotis leibii',
       'Myotis lucifugus', 'Myotis septentrionalis', 'Myotis sodalis',
       'Glaucomys sabrinus coloratus', 'Accipiter cooperii',
       'Accipiter striatus', 'Aquila chrysaetos', 'Buteo lineatus',
       'Circus cyaneus', 'Elanoides forficatus',
       'Haliaeetus leucocephalus', 'Cathartes aura', 'Pandion haliaetus',
       'Aythya affinis', 'Chlidonias niger', 'Larus atricilla',
       'Coccyzus erythropthalmus', 'Falco columbarius',
       'Falco peregrinus', 'Gavia immer', 'Grus americana',
       'Rallus elegans', 'Guiraca caerulea', 'Piranga rubra',
       'Certhia americana', 'Aimophila aestivalis',
       'Ammodramus savannarum', 'Melospiza lincolnii',
       'Loxia curvirostra', 'Progne subis', 'Riparia riparia',
       'Tachycineta bicolor', 'Lanius ludovicianus', 'Icteria virens

In [238]:
# getting to know species that have no value for the conservation_status column
species_nan = species[species.conservation_status.isna()]
species_nan = species_nan.rename(columns={'scientific_name': 'scientific', 'common_names': 'nicknames',\
                                     'conservation_status': 'status'})
species_nan.head()
species_nan.duplicated().value_counts()

False    5633
dtype: int64

#### Some analysis

Provided that the number of missing data is too high, listwise deletion is not an option in that it would reduce the sample size too much. Although, that will be required to assess the distribution of conservation status for animals.

On the other hand, in spite of the `conservation_status` value is missing, information in other columns (category, scientific name, and common names) might be important to analyse the `observation` dataset.

### 3.1.2 Handling duplicated data

Since each species has its own unique scientific name, the expected unique number should be the same of the number of rows (5824). However, there are 5541 unique values in the `scientific name`  column, which reveals the existence of duplicated entries. 



### 3.1.2.1 Inspecting `scientific_name` duplicated data

In [239]:
species1 = species.rename(columns={'scientific_name': 'scientific',  
                        'common_names': 'nicknames', 'conservation_status':'status'})

# checking duplicated rows
print("Is there duplicated rows?", species1.duplicated().unique()[0])

#checking duplicated values in the scientific column

species_dup_scientific = species1[species1.duplicated(subset=['scientific'], keep=False)].\
                     sort_values(by='scientific')

species_dup_scientific.describe(include='all')

Is there duplicated rows? False


category        scientific        nicknames              status
count              557               557              557                  23
unique               6               274              554                   4
top     Vascular Plant  Myotis lucifugus  Gray Wolf, Wolf  Species of Concern
freq               414                 3                2                  19

In [240]:
# duplicated values in the scientific_name column
species_dup_scientific[species_dup_scientific.duplicated(subset=['scientific'], keep=False)].\
   sort_values(by='scientific')

category           scientific  \
5553  Vascular Plant  Agrostis capillaris   
2132  Vascular Plant  Agrostis capillaris   
2134  Vascular Plant    Agrostis gigantea   
5554  Vascular Plant    Agrostis gigantea   
4178  Vascular Plant   Agrostis mertensii   
...              ...                  ...   
5643  Vascular Plant        Vulpia myuros   
2331  Vascular Plant     Vulpia octoflora   
4290  Vascular Plant     Vulpia octoflora   
3347  Vascular Plant         Zizia aptera   
632   Vascular Plant         Zizia aptera   

                                              nicknames status  
5553                  Colonial Bent, Colonial Bentgrass    NaN  
2132                                  Rhode Island Bent    NaN  
2134                                             Redtop    NaN  
5554                Black Bent, Redtop, Water Bentgrass    NaN  
4178               Arctic Bentgrass, Northern Bentgrass    NaN  
...                                                 ...    ...  
5643  Foxtail Fescue, Rattail Fescue, Rat-Tail Fescu...    NaN  
2331                                      Annual Fescue    NaN  
4290  Eight-Flower Six-Weeks Grass, Pullout Grass, S...    NaN  
3347  Heartleaf Alexanders, Heart-Leaf Alexanders, M...    NaN  
632                                   Golden Alexanders    NaN  

[557 rows x 4 columns]

In [241]:
# duplicated values in the scientific_name AND common_names columns
species_dup_scientific[species_dup_scientific.duplicated(subset=['scientific', 'nicknames'], keep=False)]

category           scientific        nicknames       status
3020   Mammal          Canis lupus  Gray Wolf, Wolf  In Recovery
4448   Mammal          Canis lupus  Gray Wolf, Wolf   Endangered
3283     Fish  Oncorhynchus mykiss    Rainbow Trout   Threatened
560      Fish  Oncorhynchus mykiss    Rainbow Trout          NaN

In [242]:
species_dup_scientific[species_dup_scientific.duplicated(subset=['nicknames'], keep=False)]

category                  scientific        nicknames       status
3020          Mammal                 Canis lupus  Gray Wolf, Wolf  In Recovery
4448          Mammal                 Canis lupus  Gray Wolf, Wolf   Endangered
3283            Fish         Oncorhynchus mykiss    Rainbow Trout   Threatened
560             Fish         Oncorhynchus mykiss    Rainbow Trout          NaN
1096  Vascular Plant  Silene latifolia ssp. alba  Bladder Campion          NaN
1102  Vascular Plant             Silene vulgaris  Bladder Campion          NaN

In [243]:
species_dup_scientific[(species_dup_scientific.scientific == 'Silene vulgaris') |  \
                       (species_dup_scientific.nicknames == 'Bladder Campion') | \
                      (species_dup_scientific.scientific == 'Silene latifolia ssp. alba')]

category                  scientific  \
4988  Vascular Plant  Silene latifolia ssp. alba   
1096  Vascular Plant  Silene latifolia ssp. alba   
3717  Vascular Plant             Silene vulgaris   
1102  Vascular Plant             Silene vulgaris   

                                              nicknames status  
4988  Bladder Campion, Bladder-Campion, Evening Lych...    NaN  
1096                                    Bladder Campion    NaN  
3717  Bladder Campion, Bladder Silene, Cowbell, Maid...    NaN  
1102                                    Bladder Campion    NaN

#### Some findings
Unexpectadly, among 557 rows with duplicated values in the scientific_name column, there are 274 unique values. 
It turns out that most of duplicated rows for the same `scientific_name` was due to the `common_names` values recorded. For these `scientific_nam` duplicated values, there are at least two different records for the `common_name` column. For example, there are two entries for the `scientific_name` value 'Agrostis capillaris':
- the first entry has 'Colonial Bent, Colonial Bentgrass' for the `common_name` column;
- the second entry has 'Rhode Island Bent' for the `common_name` column.

On the other hand, where there are duplicated values in the `scientific_name` column AND `common_names` column, there are different entries in the `conservation_status` column. In the 'Canis lupus' value case, there are two different values for `conservation_status` value ('In recovery' and 'Endangered').

There is a single case of `common_names` value 'Bladder Campion' which has two different entries in the `scientific_name` value: 'Silene latifolia ssp. alba' and 'Silene vulgaris'. Each one has two different entries in the `common_names` column.

### 3.1.2.2 Inspecting `common_names` duplicated data

Common names refer to the usual name of species. Conversely to scientific name, a common name often applies to multiple species.

In [244]:
#checking duplicated values in the common_names column

species_dup_nicknames = species1[species1.duplicated(subset=['nicknames'], keep=False)].\
                     sort_values(by='nicknames')

species_dup_nicknames.describe(include='all')


category           scientific           nicknames  \
count              568                  568                 568   
unique               5                  566                 248   
top     Vascular Plant  Oncorhynchus mykiss  Brachythecium Moss   
freq               330                    2                   7   

                    status  
count                   12  
unique                   4  
top     Species of Concern  
freq                     9

In [245]:
# Getting to know duplicate data in the scientific_name column AND nicknames column

species_dup_nicknames[species_dup_nicknames.duplicated(subset=['scientific'], keep=False)]

category           scientific        nicknames       status
3020   Mammal          Canis lupus  Gray Wolf, Wolf  In Recovery
4448   Mammal          Canis lupus  Gray Wolf, Wolf   Endangered
3283     Fish  Oncorhynchus mykiss    Rainbow Trout   Threatened
560      Fish  Oncorhynchus mykiss    Rainbow Trout          NaN

In [246]:
# Getting to know duplicated data in the common_names column
species_dup_nicknames

category                           scientific  \
2730  Nonvascular Plant              Dichodontium pellucidum   
2822  Nonvascular Plant               Cirriphyllum piliferum   
2022     Vascular Plant                       Carex normalis   
1971     Vascular Plant                      Carex bromoides   
1960     Vascular Plant                      Carex annectens   
...                 ...                                  ...   
250                Bird                   Dendroica coronata   
252                Bird                   Dendroica dominica   
3206               Bird                   Setophaga dominica   
2957  Nonvascular Plant  Zygodon viridissimus var. rupestris   
2956  Nonvascular Plant                 Zygodon viridissimus   

                    nicknames status  
2730                   A Moss    NaN  
2822                   A Moss    NaN  
2022           A Sedge, Sedge    NaN  
1971           A Sedge, Sedge    NaN  
1960           A Sedge, Sedge    NaN  
...                       ...    ...  
250     Yellow-Rumped Warbler    NaN  
252   Yellow-Throated Warbler    NaN  
3206  Yellow-Throated Warbler    NaN  
2957             Zygodon Moss    NaN  
2956             Zygodon Moss    NaN  

[568 rows x 4 columns]

In [247]:
species_dup_nicknames[species_dup_nicknames['nicknames'] == 'Brachythecium Moss']

category                scientific           nicknames status
2817  Nonvascular Plant   Brachythecium rutabulum  Brachythecium Moss    NaN
2814  Nonvascular Plant   Brachythecium oxycladon  Brachythecium Moss    NaN
2813  Nonvascular Plant  Brachythecium oedipodium  Brachythecium Moss    NaN
2812  Nonvascular Plant   Brachythecium digastrum  Brachythecium Moss    NaN
2816  Nonvascular Plant    Brachythecium rivulare  Brachythecium Moss    NaN
2818  Nonvascular Plant  Brachythecium salebrosum  Brachythecium Moss    NaN
2815  Nonvascular Plant    Brachythecium plumosum  Brachythecium Moss    NaN

In [248]:
# Getting to know duplicated data over than two entries in the common_names column
spe_dup_nick_over2 = species_dup_nicknames[species_dup_nicknames.duplicated(subset=['nicknames'], keep = False)]
spe_dup_nick_over2 = spe_dup_nick_over2.groupby('nicknames').count().reset_index().sort_values(by='category')
spe_dup_nick_over2.rename(columns={'category': 'counts'}, inplace=True)
spe_dup_nick_over2 = spe_dup_nick_over2[spe_dup_nick_over2.counts > 2].sort_values(by='counts', ascending=False)
spe_dup_nick_over2.describe(include='all')


nicknames     counts  scientific     status
count              41  41.000000   41.000000  41.000000
unique             41        NaN         NaN        NaN
top     Dicranum Moss        NaN         NaN        NaN
freq                1        NaN         NaN        NaN
mean              NaN   3.756098    3.756098   0.024390
std               NaN   1.157162    1.157162   0.156174
min               NaN   3.000000    3.000000   0.000000
25%               NaN   3.000000    3.000000   0.000000
50%               NaN   3.000000    3.000000   0.000000
75%               NaN   4.000000    4.000000   0.000000
max               NaN   7.000000    7.000000   1.000000

##### Some findings

There are 568 duplicated values in the `common_names` column, most of them due to the common name designates two different species. 

Among these, there are 41 `common_names` values given to more than two differente species, v.g, *'Brachythecium Moss'* common name was given to seven different species: 'Brachythecium rutabulum', 'Brachythecium oxycladon', 'Brachythecium oedipodium', 'Brachythecium digastrum', 'Brachythecium rivulare', 'Brachythecium salebrosum', and 'Brachythecium plumosum'.

There are two cases of duplicated data in the `scientific_name`column AND `common_names` column, with different `conservation_status` values.

### 3.1.2.3 Removing duplicated data

Detected duplicated data in the `scientific_name` column of `species` dataset, they must be removed in order to assess the conservation status distribution correctly.

In [249]:
# removing rows with missing values in the 'conservation_status' column

species1_nan = species1.dropna()
species1_nan['nicknames'] = species1_nan['nicknames'].apply(lambda x: 'Black-Crowned Night-Heron,' \
                                                            if 'Black-Crowned Night-Heron' in x \
                                                           else x)
species1_nan.describe(include='all')
  
    

/var/folders/mx/k35msbs11w57tlt_knr5zxbc0000gn/T/ipykernel_1896/26595849.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  species1_nan['nicknames'] = species1_nan['nicknames'].apply(lambda x: 'Black-Crowned Night-Heron,' \


category        scientific        nicknames              status
count       191               191              191                 191
unique        7               179              190                   4
top        Bird  Myotis lucifugus  Gray Wolf, Wolf  Species of Concern
freq         79                 3                2                 161

In [262]:
# split duplicated values
species1_nan_no_dup = species1_nan.drop_duplicates(subset=['scientific'])
species1_nan_no_dup

category                        scientific  \
7             Mammal                     Canis latrans   
8             Mammal                       Canis lupus   
9             Mammal                       Canis rufus   
29            Mammal                  Eptesicus fuscus   
30            Mammal         Lasionycteris noctivagans   
...              ...                               ...   
5302  Vascular Plant             Callitriche marginata   
5399  Vascular Plant  Camissonia sierrae ssp. alticola   
5426  Vascular Plant                Epilobium oreganum   
5436  Vascular Plant              Botrychium ascendens   
5676  Vascular Plant                  Romneya coulteri   

                                              nicknames              status  
7                                                Coyote  Species of Concern  
8                                             Gray Wolf          Endangered  
9                                              Red Wolf          Endangered  
29                                        Big Brown Bat  Species of Concern  
30                                    Silver-Haired Bat  Species of Concern  
...                                                 ...                 ...  
5302                              Winged Water-Starwort  Species of Concern  
5399  Mono Hot Springs Evening Primrose, Mono Hot Sp...  Species of Concern  
5426            Grants Pass Willowherb, Oregon Fireweed  Species of Concern  
5436            Trianglelobe Moonwort, Upswept Moonwort  Species of Concern  
5676                           Coulter's Matilija Poppy  Species of Concern  

[179 rows x 4 columns]

In [263]:


species1_nan_dup = species1_nan[species1_nan.duplicated(subset=['scientific'], keep=False)].\
                   sort_values(by='scientific')

species1_nan_dup

category                 scientific  \
8      Mammal                Canis lupus   
4448   Mammal                Canis lupus   
3020   Mammal                Canis lupus   
3035   Mammal           Eptesicus fuscus   
29     Mammal           Eptesicus fuscus   
172      Bird                Gavia immer   
3150     Bird                Gavia immer   
30     Mammal  Lasionycteris noctivagans   
3037   Mammal  Lasionycteris noctivagans   
3039   Mammal        Myotis californicus   
4465   Mammal        Myotis californicus   
4467   Mammal           Myotis lucifugus   
37     Mammal           Myotis lucifugus   
3042   Mammal           Myotis lucifugus   
4564     Bird      Nycticorax nycticorax   
337      Bird      Nycticorax nycticorax   
3081     Bird          Pandion haliaetus   
104      Bird          Pandion haliaetus   
3185     Bird            Riparia riparia   
226      Bird            Riparia riparia   
3029   Mammal              Taxidea taxus   
4457   Mammal              Taxidea taxus   

                                              nicknames              status  
8                                             Gray Wolf          Endangered  
4448                                    Gray Wolf, Wolf          Endangered  
3020                                    Gray Wolf, Wolf         In Recovery  
3035                       Big Brown Bat, Big Brown Bat  Species of Concern  
29                                        Big Brown Bat  Species of Concern  
172                                         Common Loon  Species of Concern  
3150  Common Loon, Great Northern Diver, Great North...  Species of Concern  
30                                    Silver-Haired Bat  Species of Concern  
3037               Silver-Haired Bat, Silver-Haired Bat  Species of Concern  
3039  California Myotis, California Myotis, Californ...  Species of Concern  
4465                                  California Myotis  Species of Concern  
4467                                Little Brown Myotis  Species of Concern  
37                Little Brown Bat, Little Brown Myotis  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
4564                          Black-Crowned Night Heron  Species of Concern  
337                          Black-Crowned Night-Heron,  Species of Concern  
3081                             Osprey, Western Osprey  Species of Concern  
104                                              Osprey  Species of Concern  
3185                          Bank Swallow, Sand Martin  Species of Concern  
226                                        Bank Swallow  Species of Concern  
3029                            American Badger, Badger  Species of Concern  
4457                                             Badger  Species of Concern

##### Some findings

Duplicated values in the 'scientific_name' column is mostly due to different entries in the 'common_names' column.

Most of the 'common_names' values for the same 'scientific_name' column is included in the other entry. For example, for *Pandion haliaetus* value in the 'scientific_name' column, there are two entries in the 'common_names' column: *Osprey* and *Osprey, Western Osprey*.

Duplicated values for *Canis lupus* and *Myotis lucifugus* will be addressed individually because there are three entries each. 

In [258]:
species1_nan_dup['more_than_one_name'] = species1_nan_dup['nicknames'].apply(lambda x: True if ',' in x else False)
species1_nan_dup = species1_nan_dup.sort_values(by='more_than_one_name')
species1_nan_dup.describe(include='all')

category   scientific        nicknames              status  \
count        22           22               22                  22   
unique        2           10               21                   3   
top      Mammal  Canis lupus  Gray Wolf, Wolf  Species of Concern   
freq         14            3                2                  19   

       more_than_one_name  
count                  22  
unique                  2  
top                  True  
freq                   12

In [259]:

species1_nan_dup.more_than_one_name.value_counts()

True     12
False    10
Name: more_than_one_name, dtype: int64

## 3.2 Inspecting the `observations` dataset

In the `observations` dataset there are three columns: `scientific_name`, `park_name`, and `observations`.


In [ ]:
observations.head(10)

In [ ]:
observations1 = observations.rename(columns={'scientific_name': 'scientific', 'park_name': 'park'})
print(observations1.info())

In [ ]:
print(observations1.describe(include='all'))

In [ ]:
print(observations1.scientific.duplicated().value_counts())

In [ ]:
print(observations1.scientific.value_counts())

In [ ]:
print(observations1.park.value_counts())

#### Some findings

There are 5541 different species spotted and 5824 entries for each park (Great Smoky Mountains National Park,
Yosemite National Park, Bryce National Park, and Yellowstone National Park).

### 3.2.1 Checking for duplicated scientific name values per park

First, data will be split per park.

In [ ]:
smoky = observations1[observations1['park'] == 'Great Smoky Mountains National Park']
smoky.sort_values(by='observations', ascending=False).head()

In [ ]:
print(smoky.scientific.duplicated().value_counts())
smoky_duplicated = smoky[smoky.duplicated(subset=['scientific'])]
smoky_duplicated[smoky_duplicated.duplicated(subset=['scientific'], keep=False)].sort_values(by='scientific')

In [ ]:
yosemite = observations1[observations1['park'] == 'Yosemite National Park']
yosemite.sort_values(by='observations', ascending=False).head()

In [ ]:
print(yosemite.scientific.duplicated().value_counts())
yosemite_duplicated = yosemite[yosemite.duplicated(subset=['scientific'])]
yosemite_duplicated[yosemite_duplicated.duplicated(subset=['scientific'], keep=False)].sort_values(by='scientific')

In [ ]:
bryce = observations1[observations1['park'] == 'Bryce National Park']
bryce.sort_values(by='observations', ascending=False).head()


In [ ]:
print(bryce.scientific.duplicated().value_counts())
bryce_duplicated = bryce[bryce.duplicated(subset=['scientific'])]
bryce_duplicated[bryce_duplicated.duplicated(subset=['scientific'], keep=False)].sort_values(by='scientific')

In [ ]:
yellowstone = observations1[observations1['park'] == 'Yellowstone National Park']
yellowstone.sort_values(by='observations', ascending=False).head()

In [ ]:
print(yellowstone.scientific.duplicated().value_counts())
yellowstone_duplicated = yellowstone[yellowstone.duplicated(subset=['scientific'])]
yellowstone_duplicated[yellowstone_duplicated.duplicated(subset=['scientific'], keep=False)].sort_values(by='scientific')

In [ ]:
# Duplicated species observations per park
smoky_dup_list = smoky_duplicated[smoky_duplicated.duplicated(subset=['scientific'], keep=False)]\
.sort_values(by='scientific', ascending=True)
smoky_dup_list.scientific.unique()

In [ ]:
yosemite_dup_list = yosemite_duplicated[yosemite_duplicated.duplicated(subset=['scientific'], keep=False)]\
.sort_values(by='scientific', ascending=True)
yosemite_dup_list.scientific.unique()

In [ ]:
bryce_dup_list = bryce_duplicated[bryce_duplicated.duplicated(subset=['scientific'], keep=False)]\
.sort_values(by='scientific', ascending=True)
bryce_dup_list.scientific.unique()

In [ ]:
yellowstone_dup_list = yellowstone_duplicated[yellowstone_duplicated.duplicated(subset=['scientific'], keep=False)]\
.sort_values(by='scientific', ascending=True)
yellowstone_dup_list.scientific.unique()

In [ ]:
species_dup_list = species_duplicated[species_duplicated.duplicated(subset=['scientific'], keep=False)]\
.sort_values(by='scientific', ascending=True)
species_dup_list.scientific.unique()

#### Some findings
There are duplicated records for the species *'Canis lupus', 'Castor canadensis', 'Columba livia', 'Holcus lanatus', 'Hypochaeris radicata', 'Myotis lucifugus','Procyon lotor', 'Puma concolor', 'Streptopelia decaocto'* for all the four parks. 

Since the same species have duplicated values, it's unlikely that happened due to random circunstances.

Those species are the same that have duplicated values in the *species* dataset. As seen above, duplicated values in the `scientific_name` column is due to different entries of `common_names` for the same species.

### 4. Distribution of conservation status for animals

Under the Endangered Species Act (ESA) at-risk species are classified as endangered species and threatened species.

An **endangered species** is any species in danger of extinction.

A **threatened species** is any species which is likely to become endangered within the foreseeable future. 

Species designated as threatened or endangered are called **listed species** in that they are added to the federal lists of endangered and threatened wildlife and plants. Species must meet the definition of endangered and threatened under the Act.

The National Park Service dataset lists at-risk species in parks and includes those under the Endangered Species Act, and also state, local, and tribal listed species.

A **species of special concern** is any species that is particularly vulnerable, and could easily become endangered.

An **in recovery species** is any species that is subjected to a recovery program in order to no longer require special protection.

##### Some findings

There are 1021 records for the category animal ('Mammal', 'Bird', 'Reptile', 'Amphibian', and 'Fish') in the dataset.

As seen below, `Bird` is the largest group, with 521 records.

In [ ]:
species_animals = species[species.category.str.contains('Plant') == False]
print('Number of entries:', species_animals.category.count())
species_animals.category.unique()

In [ ]:
ax = sns.catplot(data=species_animals, x='category', kind='count')
ax = ax.facet_axis(0,0)
for i in ax.containers:
    ax.bar_label(i,)

#### 4.1 Dealing with duplicated data

Most of duplicated rows for the same species was due to the common names recorded for the same `scientific_name` value recorded. The duplicated species are 'Canis lupus', 'Castor canadensis', 'Columba livia','Holcus lanatus', 'Hypochaeris radicata', 'Myotis lucifugus','Procyon lotor', 'Puma concolor', 'Streptopelia decaocto'.
However, in the 'Canis lupus' case, there are two different for entries in the `conservation_status` column ('In recovery' and 'Endangered').


In [ ]:
species_duplicated[species_duplicated.duplicated(subset=['scientific'], keep=False)].\
sort_values(by='scientific')


In [ ]:
print(species[species['scientific_name'] == 'Canis lupus'])

# species2 = species
# species2.rename(columns={'scientific_name': 'scientific', 'common_names': 'nicknames'}, inplace=True)
# print(species2[species2.scientific == 'Canis lupus'])
# species2['More than one common name'] = species2.apply(lambda row: True if ', ' in row['nicknames']\
#                                              else False, axis=1)
# species2.sort_values(by='More than one common name', inplace=True)
# species2[species2.scientific == 'Canis lupus']


#### 4.2 Handling missing data

As seen before, most entries in the conservation status columns (about 97%) have missing data. For 5633 records, there are 191 species with information about its conservation status in the dataset.

For the purpose of this topic, it's important to assess the distribution of missing data among animals category.


In [ ]:
# Visualizing the distribution of missing conservation status data
species_animal_null = species_animals.fillna(value='No data')
species_animal_null.head(10)

order=['Species of Concern','Threatened', 'Endangered', 'In Recovery', "No data"]

plt.figure(figsize=(8,6))
ax = sns.catplot(data=species_animal_null, x='conservation_status', kind='count', \
                 order= order)
ax = ax.facet_axis(0,0)
for i in ax.containers:
    ax.bar_label(i,)
axe = plt.subplot()
axe.set_xticks(ticks=range(5))
axe.set_xlabel(order)
axe.set_xticklabels(labels = order, rotation=45)
plt.title("Distribution of conservation status for animal including null values")
plt.show()
plt.clf()


##### Some findings

Missing data outnumber other `conservation_status` values for all groups of animals as seen on the graphs below. 

Possilble causes for that are:
- most species have not been included on the list;
- those species' conservation status were not informed.

In [ ]:
g = sns.catplot(data=species_animal_null, x='conservation_status', col='category', col_wrap = 5,\
            kind='count', height=3, aspect=8/5, \
            order=['Species of Concern','Threatened', 'Endangered', 'In Recovery', "No data"])

g.fig.tight_layout()
g.set_xticklabels(labels=order, rotation=45)

plt.show()
plt.clf()

In [ ]:
g = sns.catplot(data=species_animal_null, x='category', hue='conservation_status',\
            kind='count', height=4, aspect=13/5, legend_out=False)

sns.move_legend(g, "upper right")
plt.title("Distribution of missing values among animals")
plt.show()
plt.clf()

In [ ]:
# Visualizing the distribution of conservation status

plt.figure(figsize=(10,7))
ax = sns.catplot(data=species_animals, x='conservation_status', kind='count', \
                 order=['Species of Concern','Threatened', 'Endangered', 'In Recovery'])
ax.set_xticklabels(rotation=45)
plt.title("Distribution of conservation status for animals")
plt.show()
plt.clf()

In [ ]:
animals = sns.catplot(data=species_animals, x='conservation_status', col='category', col_wrap = 2,\
                      kind='count', order=['Species of Concern','Threatened', 'Endangered', 'In Recovery'],\
                      height=3, aspect=13/6)
plt.show()
plt.clf()


#### Some findings

Most animals were listed as *species of concern* conservation status.



## Which species were spotted the most at each park?

## Selecting species with conservation status data

In [ ]:
# merging

conservation = species.merge(observations, on='scientific_name')
conservation.head()
conservation.info()
print(conservation.duplicated(subset=['scientific_name']).value_counts())
conservation['conservation_status'].isna().value_counts()

In [ ]:
conservation1 = conservation[conservation['conservation_status'].notnull()]

In [ ]:
conservation1.info()

In [ ]:
conservation1.duplicated(subset=['scientific_name']).value_counts()

In [ ]:
conservation1.head(10)

In [ ]:
conservation1.groupby('scientific_name').sum('observations')

- What is the distribution of `conservation_status` for animals?

- Are the differences between species and their conservation status significant?

- Are certain types of species more likely to be endangered?